internal/sharedcomponent/sharedcomponent.go

```diff
@@ -0,0 +1,78 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sharedcomponent
+
+import (
+	"context"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+type SharedComponents struct {
+	comps map[interface{}]*SharedComponent
+}
+
+func NewSharedComponents() *SharedComponents {
+	return &SharedComponents{
+		comps: make(map[interface{}]*SharedComponent),
+	}
+}
+
+func (scs *SharedComponents) GetOrAdd(key interface{}, create func() component.Component) *SharedComponent {
```

Comment:
Comment? Maybe package comment too would be nice for high level overview.

Reply:
Done. Commented all funcs and package